packages/medusa/src/utils/transaction/transaction-builder.ts

```diff
@@ -0,0 +1,358 @@
+  depth: number
+  parent?: InternalStep | null
+}
+
+export class TransactionBuilder {
+  private steps: InternalStep
+
+  constructor(steps?: TransactionStepsDefinition) {
+    this.load(steps)
+  }
+
+  load(steps?: TransactionStepsDefinition) {
+    this.steps = {
+      depth: -1,
+      parent: null,
+      next: steps
+        ? JSON.parse(
+            JSON.stringify((steps.action ? steps : steps.next) as InternalStep)
+          )
+        : undefined,
+    }
+
+    this.updateDepths(this.steps, {}, 1, -1)
+    return this
+  }
+
+  addAction(
+    action: string,
+    handler: ActionHandler,
+    options: Partial<TransactionStepsDefinition> = {}
+  ) {
+    let step = this.steps as InternalStep
+
+    while (step.next) {
+      step = Array.isArray(step.next)
+        ? (step.next[step.next.length - 1] as InternalStep)
+        : (step.next as InternalStep)
+    }
+    const newAction = {
+      action,
+      depth: step.depth + 1,
+      parent: step.action,
+      ...options,
+    } as InternalStep
+
+    step.next = newAction
+
+    return this
+  }
+
+  replaceAction(
+    existingAction: string,
+    action: string,
+    handler: ActionHandler,
+    options: Partial<TransactionStepsDefinition> = {}
+  ) {
+    const step = this.findStepByAction(existingAction)!
+    step.action = action
+
+    Object.assign(step, options)
+
+    return this
+  }
+
+  insertActionBefore(
+    existingAction: string,
+    action: string,
+    handler: ActionHandler,
+    options: Partial<TransactionStepsDefinition> = {}
+  ) {
+    const parentStep = this.findParentStepByAction(existingAction)
+    if (parentStep) {
+      const oldNext = parentStep.next!
+      const newDepth = parentStep.depth + 1
+      if (Array.isArray(parentStep.next)) {
+        const index = parentStep.next.findIndex(
+          (step) => step.action === existingAction
+        )
+        if (index > -1) {
+          parentStep.next[index] = {
+            action,
+            ...options,
+            next: oldNext[index],
+            depth: newDepth,
+          } as InternalStep
+        }
+      } else {
+        parentStep.next = {
+          action,
+          ...options,
+          next: oldNext,
+          depth: newDepth,
+        } as InternalStep
+      }
+
+      this.updateDepths(oldNext as InternalStep, parentStep)
+    }
+
+    return this
+  }
+
+  insertActionAfter(
+    existingAction: string,
+    action: string,
+    handler: ActionHandler,
+    options: Partial<TransactionStepsDefinition> = {}
+  ) {
+    const step = this.findStepByAction(existingAction)!
+    const oldNext = step.next
+    const newDepth = step.depth + 1
+    step.next = {
+      action,
+      ...options,
+      next: oldNext,
+      depth: newDepth,
+      parent: step.action,
+    } as InternalStep
+
+    this.updateDepths(oldNext as InternalStep, step.next)
+
+    return this
+  }
+
+  moveAction(
+    actionToMove: string,
+    targetAction: string,
+    runInParallel = false
```

Comment:
as mention earlier, should we have an options object to be more explicit and to be able to add more options if necessary later

Reply:
I will transform it in an object, even though this method will be renamed and turn private so `move` and `merge` will be aliases for it.

Reply:
👍